### PR TITLE
ENG-94839: report actions whose target page does not exist on mobile

### PR DESCRIPTION
**Pull request by MaksymKryvonos**
[ENG-94839](https://creatio.atlassian.net/browse/ENG-94839) · epic [ENG-93494](https://creatio.atlassian.net/browse/ENG-93494) *Converting Web page to Mobile page*

## Problem

`get-mobile-page-conversion-guide` checks whether an action's **request type** exists on mobile. It never checked where the request **goes**.

`crt.OpenPageRequest` carries `params.schemaName`; `crt.CreateRecordRequest` / `crt.UpdateRecordRequest` carry `params.entityName`. All three are `DirectMapping`, so they convert cleanly — and the converted page ships a control that navigates to a page which does not exist on mobile. The developer finds out at runtime, in the app.

Real example, from the pinned fixture `LeadsFormPage.live-snapshot.json`: `PostponeQueueItemButton` → `crt.OpenPageRequest{schemaName: "PostponeQueueItemPage"}`, plus three `crt.CreateRecordRequest` buttons for `LeadProduct` / `Lead` / `Opportunity`. Four actions, no target check at all today.

## What this does

Every unreachable target is reported as typed data on `requestConversions`. **The control always survives.** What can be lost is the ACTION, and only when its absence is *definitional*:

```jsonc
"unresolvedTargetRequests": [
  { "elementName": "PostponeQueueItemButton", "binding": "clicked",
    "webRequest": "crt.OpenPageRequest", "targetKind": "web-page",
    "target": "PostponeQueueItemPage", "state": "missing",
    "bindingRemoved": true },
  { "elementName": "ProductsAddButton", "binding": "clicked",
    "webRequest": "crt.CreateRecordRequest", "targetKind": "entity-default-mobile-page",
    "target": "LeadProduct", "state": "missing",
    "bindingRemoved": false }
],
"targetsProbed": true
```

| kind | verdict comes from | on `missing` |
|---|---|---|
| `web-page` | nothing — a web page cannot open on mobile **by construction** | binding removed, control kept |
| `entity-default-mobile-page` | reading the object's `MobileRelatedPage` add-on | **reported only, nothing touched** |

`bindingRemoved` is the field to branch on; `state` says only how confidently the target is reported. Matching **constraints** carry the instruction, because the guide's contract is to be self-describing, plus `targetsNote` when the check was incomplete — so an empty list is never misread as "all clear".

> Earlier revisions of this PR removed the action for *both* kinds. That was cut back after review: an object's verdict is a read that cannot prove absence, so it never costs an action. Removal for that kind, if ever wanted, is the scope of ENG-96178 / ENG-95084, and `MobileActionTargetProbe.StripsBindingOnMissing` is the one line where it would hook in.

## Design notes worth a reviewer's attention

**Which requests carry a target is data, not code.** A rule's `targetParam` / `targetKind` (`WebToMobilePageConversionRules.json`). `crt.ImportDataRequest`, `crt.AddNextStepRequest` and friends can be covered by a CDN rules push with no clio release, and an **unrecognized** kind degrades to silence — a newer rules file can never make an older clio report a target it cannot verify. It also cannot make an older clio *remove* anything: the strip seam is keyed on a kind constant, not on rules text.

**Only a verdict that needed no environment call may remove an action.** That is the whole shape of the feature. A `web-page` target is settled offline, and the invariant it rests on is that a WEB page's `schemaName` names a web page — which is why the two-tier `SysSchema`/`GetParentSchemas` classifier originally built to *verify* it was deleted: the answer was fixed. The object tier reads the object's `MobileRelatedPage` add-on with a cheaper request shape than `RelatedPageAddonService` uses, and its `missing` is deliberately non-destructive.

**The object tier's read is verified against a stand.** See the results comment below: the probe's exact shape agrees with the canonical read on both branches, and `TargetPackageUId` does not affect the answer under `UseFullHierarchy`. The `web-page`-only strip rule is therefore a deliberately conservative policy, not a workaround for an unverified read.

**Fail-open is structural, not a flag.** An absent resolution reads as `Unknown`, so the empty-dictionary default *is* the safe path — there is no `if (probeOk)` in the analysis service. None of these can conclude absence: an unreachable environment, a proxy/auth error page, a `SelectQuery` returned at its row cap, an exhausted read ceiling, a blank add-on body, or **the object the page itself is bound to** (its mobile page is this conversion's own closing step — those are dropped from the report entirely).

**The probe degrades PER TIER.** `ProbeOk` reports the tier that needs the environment. Collapsing both tiers into one boolean meant a page carrying one web-page target *and* one object target silently lost the web-page warning that the same page without the object target reports fine — which is exactly the real `Leads_FormPage` shape.

**A finding can never name a control the mobile page does not have.** It is emitted where the binding is written, and swept if a later removal pass drops its element. The removal decision additionally depends on the WRITER: on a same-component twin, `mobileValues` is a delta *merge* payload where an omitted key means "keep the mobile template's own value", so a removal is not something that writer can perform.

**A row cap on an unordered read cannot be trusted with the row you need.** The object read caps `SysSchema` at a few rows per name, and a `SelectQuery` applies `rowCount` to an *unordered* result. On the stand, six common objects each carry 8 schema layers against a 4-row window, and 4 of the 6 lost their base row to it — degrading a real verdict to `unknown`. The read now orders `ExtendParent` ascending, which puts the base row first by construction.

## Review

Pre-PR gate run as three parallel agents (correctness / security / quality) over the full diff. `dkorotin1`'s review at `f6ff0a413` is answered in the thread; the second round found a twin-merge writer asymmetry, an unusable-package path that claimed a verification it never ran, and that plain `Redact` is the wrong primitive for server-authored prose crossing the MCP boundary — all fixed in `f686e3d07`.

`Validated: dotnet test --filter "Category=Unit&(Module=McpServer|Module=Command)"` — 8804 passed, 0 failed. (`IdentityServiceDeployment` is excluded: its 5 failures reproduce unchanged on the parent commit — a macOS temp path is a filesystem reparse point.) Build clean, no new `CLIO*` warnings.

**MCP reviewed** — response gains `unresolvedTargetRequests` (with `bindingRemoved`) / `targetsProbed` / `targetsNote`; no argument, tool name or destructive flag changed. This tool has no curated `get-tool-contract` entry, no MCP prompt or resource, and no `McpCapabilityMap` section, so nothing else to align. E2E extended in `MobilePageConversionGuideSandboxE2ETests`. **Docs reviewed, no update required** — MCP-only, no CLI verb, no `help/en` and no `docs/commands` page. **ClioRing compatibility reviewed, no Ring-consumed contract changed** — Ring calls only `clio-run`, `get-tool-contract`, `list-environments` and `deploy-creatio` (inspected `clio-ring/ClioRing.Ipc`, `clio-ring/ClioRing`, `clio-ring/ClioRing.Desktop/actions.json`); this tool is feature-gated, MCP-only and non-resident.

## Before merging

1. ~~Stand check on the add-on request shape.~~ **Done** — verified on DevMK, both branches; see the results comment.
2. **`clio-knowledge` follow-up.** The article `freedom-page-web-to-mobile-conversion` should document `unresolvedTargetRequests` / `bindingRemoved` / `targetsProbed` / `targetsNote`, and that the converter now removes a binding for a `web-page` target. It ships from a separate repository and needs its own PR plus a `libraryVersion`/`sequence` bump. Nothing there is stale today — the article does not mention this surface yet, and the guide's own constraints carry the instruction in the meantime.

### Conflict note

PR #1365 (ENG-95827) rewrites the same converter files. This change was deliberately shaped to merge mechanically with it: `RequestConversionInfo` is byte-identical on both branches, everything environment-facing is a new file, and no `ElementMapContext`, drop-gate or `ClassifyClicked` change was needed. `ProcessOneEventBinding` does gain a parameter.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[ENG-94839]: https://creatio.atlassian.net/browse/ENG-94839?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ENG-93494]: https://creatio.atlassian.net/browse/ENG-93494?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by dkorotin1 — CHANGES_REQUESTED** [review]
Reviewed at `f6ff0a413` against ENG-94839. The design is sound — fail-open is structural (`ActionTargetState.Unknown = 0`), `targetParam`/`targetKind` live in the rules file with silent degradation on an unknown kind, and the unit/E2E coverage of the degraded paths is genuinely thorough. Two things block, and both come from the last commit turning this from warn-only into something that removes a binding.

## Blocker 1 — the `MobileRelatedPage` read shape is unverified, and a wrong answer now deletes a working action

`MobileActionTargetProbe.cs:458-478` addresses the add-on with `TargetParentSchemaUId = Guid.Empty` and `TargetPackageUId` = the **source page's** package. The canonical read (`RelatedPageAddonService.cs:591-600`) passes `entitySchema.ParentSchema?.UId` and the **object's own** package. `GetSchema` auto-provisions the descriptor (`RelatedPageAddonService.cs:462`), so a mis-addressed read plausibly returns a fresh `{}` / `{"Pages":[]}` body — which is not blank, so `ClassifyRelatedPageMetadata` returns `Missing` (`:508-509`) and `ProcessOneEventBinding` strips the binding. Every `crt.CreateRecordRequest` / `crt.UpdateRecordRequest` on the page then silently loses its action.

The stand evidence in `f6ff0a413` only exercises the negative branch (three objects reported missing). The positive branch — an object that *does* have a default mobile page resolving to `Resolved` — has not been observed anywhere, and it is the branch this depends on.

Fix: verify that exact request shape on a stand against an object with a configured default mobile page (and pin it as a knowledge record), or align the request with `BuildAddonGetRequest`, or drop `targetKind` from the two record requests as the PR itself offers.

## Blocker 2 — the web-page verdict is discarded on every degraded path

Web-page targets are settled as `Missing` before the environment is touched (`MobileActionTargetProbe.cs:140-143`), but `NotProbed` (`:544-546`) does not carry `TargetsByKey`, and `Analyze` gates on `ProbeOk: true`. So:

| Page | Environment | Result |
|---|---|---|
| web-page targets only | unreachable | warning reported (`:147-149`) |
| web-page **+** entity targets | unreachable | web-page warning **silently lost** |

The real `Leads_FormPage` is the second row (1 web-page + 3 entity targets), so an unreachable environment drops `PostponeQueueItemButton → PostponeQueueItemPage` — the example the ticket itself names.

Fix: split "probed" per tier — carry the already-settled web-page resolutions through the degraded paths and let `ProbeOk` mean "the entity tier answered".

## Major — the PR description no longer describes the change

The body says "**The control always stays on the page.** … Neither removes anything" and defers removal to ENG-96178 / ENG-95084. `f6ff0a413` strips the binding on `state=missing`. Same drift in `MobilePageConversionGuideTool.cs:208-211` ("changes no conversion decision").

This also needs a call from the reporter: the ticket says "warn user but keep this button/menu item on page". The control is kept, the action is not — that is a behavioural decision beyond the AC, and the PR previously said it belonged to another ticket.

Related: `crt.OpenPageRequest` bindings are now stripped unconditionally, offline runs included, with no environment check. The "a web `schemaName` names a web page by construction" assumption is defensible (client-unit schema names are unique per manager) but it is now load-bearing for a destructive edit rather than a warning, and it is exactly what the deleted two-tier classifier was built to test. At minimum gate the strip on `ProbeOk`, and record the naming-uniqueness fact.

## Minor

- Dangling crefs from the deleted classifier: `MobileActionTargetProbe.cs:74` (`MaxAuthoritativeProbes`), `:418` (`SchemaLookup{TRow}`), and two stacked `<summary>` blocks on `ResolveEntityTargets` (`:360-372`).
- Vocabulary drift in the public contract: `UnresolvedTargetRequest.TargetKind`, `ActionTargetResolution.Kind` and `RequestMappingRule.TargetKind` all document `mobile-page`; the rules file and the code emit `web-page`. A consumer branching on the documented literal matches nothing.
- Truncated sentence in `RequestConversionInfo.UnresolvedTargetRequests`: "The action is (ENG-94839)."
- `MobileActionTargetProbe.cs:131` interpolates a raw `ex.Message` into `Note`, which reaches the caller as `targetsNote`; the sibling path at `:178` redacts. Low risk, but the two should match.
- The E2E asserts against the wrong field: `survivingWebNames` is built from `e.WebName`, while `finding.ElementName` is the mobile name (`WebToMobileAnalysisService.cs:3181` passes `mobileName`). It passes only because names usually coincide, and will false-fail on a renamed element.
- `docs/knowledge/.../client-unit-schema-type-is-not-a-sysschema-column.md` is worth keeping, but its `applies-to` points at files this PR does not touch, so `check-knowledge` will never surface it to whoever edits the code it came from. Meanwhile the two facts this change actually rests on (the add-on request shape; schema-name uniqueness) have no record.
- `ClassifyEntityDefaultMobilePage:464` resolves `IAddonSchemaDesignerClient` inside the loop (up to 8 times) — hoist it into `ProbeContext`. And when the 8-probe budget is exhausted the remaining targets become `Unknown` with no note, so the caller cannot tell "not asked" from "asked, no answer".

## Test run

`dotnet test clio.tests/clio.tests.csproj --filter "Category=Unit&Module=McpServer"` at `f6ff0a413`: **Failed: 0, Passed: 4738, Skipped: 2** (4 m 31 s, net10.0) — consistent with the description. (An earlier run of the same filter reported one failure that did not reproduce and that I did not capture; treating it as flaky/order-dependent, not a regression.)

**Comment by MaksymKryvonos** [comment]
@dkorotin1 — validated every item against the code at `f6ff0a413`; the review holds almost entirely. Addressed in `f686e3d07`. Both blockers had the same root cause you named, and the fix follows the reporter's call on scope.

## Blocker 1 — confirmed, and stronger than written

Your citations are exact. `{}` → `Missing` is not merely plausible, it is **pinned as intended** by this PR's own `[TestCase("{}", ActionTargetState.Missing, TestName = "ClassifyRelatedPageMetadata_NoPagesKey_IsMissing")]`. And `ClassifyRelatedPageMetadata`'s own remarks already acknowledged the shortcut ("a shortcut must never present as a verified absence") while guarding only the *blank* shape — not the `{}` one the shortcut actually produces.

Resolved by option (c), scoped: **only a kind whose absence is definitional may remove an action.** New single seam `MobileActionTargetProbe.StripsBindingOnMissing(kind)`, true for `web-page` only. `crt.CreateRecordRequest` / `crt.UpdateRecordRequest` are now reported and left completely untouched — the request is not modified at all. So a wrong add-on read costs a spurious warning, which is the risk level this PR was reviewed at before that commit. Recorded in `docs/knowledge/McpServer/mobile-related-page-addon-read-shape-is-unverified.md`, with the stand check that deletes it.

## Blocker 2 — confirmed executably

Reproduced your table with three throwaway tests before fixing. Root cause exactly as you described: `NotProbed` (`:544-546`) drops `TargetsByKey`, and `Analyze` gated on `ProbeOk`. Worth noting **every** existing degraded-path test used a single-tier fixture — `Probe_SelectQueryFails_DegradesWithoutThrowing` and `Probe_WithoutResolver_ReportsNotProbed` carry only an entity target, `Probe_WebPageTargetWithoutResolver_StillReportsIt` only a web-page one. No fixture mixed the two, which is what hid it, and `Probe_SelectQueryFails`'s description ("degrades the **whole** probe") had codified it as intended.

Fixed your way — split per tier: `NotProbed` carries what was settled without a read, `ProbeOk` now means "the object tier answered", and the `ProbeOk` gate is gone from `Analyze` (an absent resolution is already the fail-open signal — which is what the design said before that gate was added). Two mixed-tier tests now pin it.

One correction: **`ProbeOk` was not the missing gate you asked for in the Major** — `WebToMobileAnalysisService.cs:209` already had it. It just did nothing, because `ProbeOk` is true for a web-page-only page without any environment read. That ask also pulled against Blocker 2, which wants the web-page verdict to *survive* a degraded probe. They only reconcile by separating "report" from "strip", which is what the kind seam does.

## Major — confirmed; PR body rewritten

You were right that the body described a different change. Rewritten to match. Two additions to your list: the E2E's own `because` still said "the finding is a **warning**", and `MobilePageConversionGuideTool.cs:208-211` was scoped to "an unreachable environment", so it was accurate for the mixed case and wrong only for the web-page-only one — both fixed.

New typed field so the caller never has to re-derive the rule: **`unresolvedTargetRequests[].bindingRemoved`**. Branch on it for what happened; `state` now says only how confidently the target is reported.

## Minor — all confirmed, all fixed

- **Dangling crefs.** All three. Also `MobileActionTargetProbe.cs:117`'s comment cited the deleted `CarriesBindingOnMobile`, and `ResolveEntityUId` was ambiguous across two overloads. Note none of these warn: `clio.csproj` sets no `GenerateDocumentationFile`, so CS1574/CS1571 never fire — they are silent.
- **`mobile-page` vs `web-page`.** Three doc sites, and — more importantly — **the tests carried the same stale literal**, so their fixtures never exercised a kind clio emits. All bound to `MobileActionTargetProbe.KindWebPage` now. Your finding surfaced this: the fixtures only started failing once the kind became load-bearing.
- **Truncated sentence.** Fixed.
- **Raw `ex.Message` at `:131`.** Fixed — and the security pass escalated it: both notes carry prose the *server* chooses (`DataServiceSelectResponse` interpolates `errorInfo.message` verbatim), so plain `Redact` was the wrong primitive — it scrubs values, not prose: no flattening, no clamp, no fence. Both paths now use `RedactUntrustedOrNull`, per the rule already written on `CreatioResponseError.DescribeServerReportedReadError`. Three cases pin it (credential pair, tenant URI, forged instruction block).
- **E2E asserting the wrong field.** Confirmed: `ElementName` is the mobile name (`ProcessEventBindings(ctx, node, values, mobileName)`). Also worse than described — `WebName` is legitimately `null` for an inserted element (the E2E's own `:451` asserts that), and those get filtered out of the set. Keyed on `MobileName` now, plus an assertion that `bindingRemoved` implies `state: missing` and a matching `droppedRequests` entry.
- **Knowledge records.** Half agreed: the existing record's `applies-to` paths exist and match its *subject* (`list-pages`/`PageSchemaType`), so pointing there is defensible — I added the probe so it surfaces. Your second half was right and is done: two new records for the facts this change rests on, plus a third for something the second round found.
- **Client resolved in the loop + silent budget exhaustion.** Both fixed — hoisted into `ProbeContext` (with a test that it resolves once), and the ceiling now sets a note. `targetsNote` is therefore surfaced even when `targetsProbed` is true, since "ran but incomplete" is not expressible by the flag alone.

## Second round found three more in the same surface

- **`ProcessOneEventBinding` is shared by two writers with opposite omission semantics.** On the insert path a missing key removes the property; on the same-component twin path `values` is a delta *merge* payload where a missing key means "keep the mobile template's own value". It was reporting `bindingRemoved: true` **plus** a `droppedRequests` entry for a merge that removes nothing — and telling the agent the control "renders and does nothing" when it may still fire the template's request. Now gated per writer; pinned by a test verified to fail without the fix.
- **An unusable `PagePackageUId` reported `targetsProbed: true` with no note after zero reads** — contradicting both contracts. The object tier now returns whether it *answered*, matching what `PageBusinessRuleProbe` already does for the same input.
- **The three constraint groups were only exclusive by an invariant held elsewhere**, and hard-coded one kind's reason. Now exclusive by construction, with the remedy composed from the kinds actually present — so admitting a second kind to the strip seam cannot make them lie.

Plus contract docs realigned where they still described the old behaviour: the probe-result class summary promised an EMPTY map on any failure, `unresolvedTargetRequests` promised emptiness when unprobed, `ReclassifyRemovedTargetFindings` carried a `<summary>` copied from another method, and the `ActionTargetState` docs named a deleted escalation tier. `Analyze_MenuItemTargetMissing` also used a plain `crt.Button`, so it pinned nothing about menu items — it builds a real `menuItems` child now.

## Test run

`dotnet test clio.tests/clio.tests.csproj --filter "Category=Unit&Module=McpServer"` at `f686e3d07`: **Failed: 0, Passed: 4751, Skipped: 2**. Build clean, no new `CLIO*` warnings.

Your before-merge stand check is still listed, but downgraded to non-blocking for the reason above.

**Comment by MaksymKryvonos** [comment]
## Stand results — Blocker 1 verified, and it does not hold

@dkorotin1 your Blocker 1 was the right call on the evidence available: the positive branch had never been observed, and until it was, a wrong answer removed a working action. It is now verified on a live stand (DevMK, `10.48.14.89:1025`), and **the read is correct**.

I sent the probe's *exact* request shape by hand to `ServiceModel/AddonSchemaDesignerService.svc/GetSchema` — base-row `UId` from `SysSchema`, `TargetParentSchemaUId = 00000000-…`, and deliberately a **foreign** `TargetPackageUId` (`UsrWebToMobileTest`, not the object's own `Custom`) — and compared it against the canonical read (`clio get-related-page-addon --schema-type mobile`, which goes through `RelatedPageAddonService.BuildAddonGetRequest`).

| Object | canonical read | probe's cheap shape | probe verdict |
|---|---|---|---|
| `Contact` | `Contact_MobileFormPage`, `isDefault` | `Pages:[{PageSchemaUId: 868f21b3-…, IsDefault: true, TypeColumnValue: null}]` | **`Resolved`** ✅ |
| `Opportunity` | `pageCount: 0` | `{"Pages":[],"TypeColumnUId":null}` | **`Missing`** ✅ |
| `LeadProduct` | `pageCount: 0` | `{"Pages":[]}` | **`Missing`** ✅ |
| `Activity` | `pageCount: 0` | `{"Pages":[]}` | **`Missing`** ✅ |

Both branches agree with the canonical read. Two specifics worth recording:

- **`TargetPackageUId` does not affect the answer** under `UseFullHierarchy: true`. `Contact` read through `Custom`, `UsrWebToMobileTest`, `UsrZen` and `MobileAutoTest` returns an identical page set and the same `pageSchemaUId`, even though the resolved *entity schema* `UId` differs per package (that is the per-package layer). So the "source page's package" shortcut is sound.
- **The feared `{}` did not appear.** An object with no configuration returns a well-formed `{"Pages":[],"TypeColumnUId":null}`, which classifies as `Missing` — correctly. So the auto-provisioned descriptor is not the ambiguous shape the risk assumed.

This also confirms the negative evidence in `f6ff0a413` was right: `LeadProduct` and `Opportunity` genuinely have no default mobile page.

**What does not change.** The strip stays scoped to `web-page` only. The reporter's call was that a `CreateRecordRequest` / `UpdateRecordRequest` must not be touched, and that stands — it is now a deliberately conservative policy rather than a workaround for an unverified read.

## A different, real defect found in the same session

While pulling the `SysSchema` rows I hit something worse than the add-on question.

`ReadEntitySchemaRows` caps its read at `RowsPerNameHeadroom` (4) rows **per name**, and a `SelectQuery` applies `rowCount` to an **unordered** result. An OOTB object routinely carries more layers than that — on the stand `Contact`, `Opportunity`, `Activity`, `Account`, `Lead` and `Case` have **8** `SysSchema` rows each. Whether the base row (`ExtendParent = false`) landed inside the 4-row window was luck of the DB's ordering:

| Object | rows | base row inside the 4-row window |
|---|---|---|
| `Contact` | 8 | yes (index 1) |
| `Opportunity` | 8 | yes (index 2) |
| `Order` | — | yes (index 1) |
| `Account` | 8 | **cut off** |
| `Lead` | 8 | **cut off** |
| `Activity` | 8 | **cut off** |
| `Case` | 8 | **cut off** |

It fails open — `seenNames.Contains(name) || truncated` sends it to `Unknown`, never a false `Missing` — so it is not a correctness bug. But four of seven common objects reported "could not verify" instead of an answer, which means the object tier largely did not work on a real stand.

Fixed by ordering `ExtendParent` ascending (`false` sorts before `true`), so the base row is first by construction rather than by luck. Re-verified on the stand: `Account`, `Lead`, `Activity` and `Case` all return their base row at index 0 with the same 4-row window. The headroom keeps only its remaining job — seeing enough extra layers to separate "no rows at all" from "rows but no base row". Pinned by `Probe_EntitySchemaRead_OrdersBaseRowsFirst`.

## One thing you asked for that is not in the branch

You asked for knowledge records for the two facts this change rests on. They were written, and then **removed at the author's request** — including `client-unit-schema-type-is-not-a-sysschema-column.md`, which this PR had added earlier. Flagging it explicitly so the disappearance is not read as an oversight: it departs from the `AGENTS.md` knowledge-base policy, and it is the author's call, not a miss. The facts themselves are documented in code — `MobileActionTargetProbe.StripsBindingOnMissing` carries the add-on-shape reasoning, and `ColumnOrderedAsc` carries the row-cap one.

## Test run

`dotnet test clio.tests/clio.tests.csproj --filter "Category=Unit&(Module=McpServer|Module=Command)"` at `f2e9c2e50`: **Failed: 0, Passed: 8804, Skipped: 29**.

`IdentityServiceDeployment` is excluded from that filter: its 5 failures reproduce unchanged with my changes stashed on the parent commit — `EnsurePathHasNoReparsePoints` rejects a macOS temp path because `/var` is a symlink. Unrelated to this PR, and pre-existing.

File: clio/Command/McpServer/Data/WebToMobilePageConversionRules.json [hunk 1]
```diff
@@ -207,13 +207,13 @@
...
-    { "web": "crt.UpdateRecordRequest", "mobile": "crt.UpdateRecordRequest", "category": "DirectMapping" },
+    { "web": "crt.CreateRecordRequest", "mobile": "crt.CreateRecordRequest", "category": "DirectMapping", "targetParam": "entityName", "targetKind": "entity-default-mobile-page" },
+    { "web": "crt.UpdateRecordRequest", "mobile": "crt.UpdateRecordRequest", "category": "DirectMapping", "targetParam": "entityName", "targetKind": "entity-default-mobile-page" },
     { "web": "crt.DeleteRecordRequest", "mobile": "crt.DeleteRecordRequest", "category": "DirectMapping" },
     { "web": "crt.CopyRecordRequest", "mobile": "crt.CopyRecordRequest", "category": "DirectMapping" },
     { "web": "crt.CancelRecordChangesRequest", "mobile": "crt.CancelRecordChangesRequest", "category": "DirectMapping" },
     { "web": "crt.ClosePageRequest", "mobile": "crt.ClosePageRequest", "category": "DirectMapping" },
-    { "web": "crt.OpenPageRequest", "mobile": "crt.OpenPageRequest", "category": "DirectMapping" },
+    { "web": "crt.OpenPageRequest", "mobile": "crt.OpenPageRequest", "category": "DirectMapping", "targetParam": "schemaName", "targetKind": "web-page" },
```
Comment (MaksymKryvonos) [inline clio/Command/McpServer/Data/WebToMobilePageConversionRules.json:216]: Maybe "category": "RequiresManualDecision" ?

**Review by sokolovskyiCreatio — APPROVED** [review]
Approve with minor issues. All three ACs are implemented and covered by some tests, but the diff has three medium-confidence Major gaps worth addressing: the twin-merge path can leave a definitionally-dead crt.OpenPageRequest binding active despite AC-2's disable requirement; the real-page regression test nulls ModelConfig so AC-1's primary-object exemption is never actually exercised against real Leads_FormPage data; and the addon request-shape fix from the prior review round still lacks a test pinning the exact request DTO or a live positive-Resolved round trip. Remaining items are Minor test-design/edge-case notes. No Blocker or high-confidence Major exists, so nothing blocks merge. Two knowledge-base findings and one performance finding were dropped for failing the AC-anchor gate (decorative bindings, not objective bugs); one finding duplicated an existing open PR thread.

**Requirement Coverage**

AC-1 (create/update record request: check if default page exists) - covered; MobileActionTargetProbe resolves entity-default-mobile-page targets and reports Missing/Unknown without removing the binding/control, matching "keep this button/menu item." Adequacy gaps remain: the addon request-shape/positive-branch fix is unpinned by tests, the real-page regression test doesn't actually exercise the primary-object exemption, classification tests bypass the public API, and one classification edge case (non-array Pages shape) defaults to Missing instead of Unknown.

AC-2 (crt.OpenPageRequest: disable request) - covered for the primary path (web-page targets unconditionally resolve Missing offline and strip the binding), but the twin-merge path can leave a provably-dead OpenPageRequest binding active and mislabels it as unproven, a correctness gap against this AC.

AC-3 (warning message for unconverted-page actions) - covered via structured unresolvedTargetRequests data and free-text guide constraints; the one E2E test asserting this end-to-end can pass vacuously if no qualifying seeded page is found.

**Findings (summary only)**

- Major - Twin-merge path can leave a definitionally-dead crt.OpenPageRequest binding active while telling the caller it's unproven (`clio/Command/McpServer/Tools/MobilePageConverter/WebToMobileAnalysisService.cs:2860`). In `BuildDeltaTwinMergeValues`, `ProcessOneEventBinding` is called with `canRemoveBinding: false` whenever an event binding differs from the twin baseline. For a `crt.OpenPageRequest` naming a web page, `StripsBindingOnMissing` would normally return true, but `removed = missing && canRemoveBinding && StripsBindingOnMissing(...)` forces `removed=false` purely because it is a merge payload. The finding is still recorded with `State=Missing, BindingRemoved=false`, routing it into the "reportedMissing" guidance bucket whose text says the absence is not proven and the user should decide, which is factually wrong for `KindWebPage`, where no read was involved and the absence is definitional. Net effect: a page element with a component twin and a customized OpenPageRequest binding to a web page keeps an action that will never work, contrary to AC-2's disable-request requirement, and the tool's own guidance mischaracterizes it as discretionary rather than definite. Fix: in `AddUnresolvedTargetConstraints`, branch the "reportedMissing" bucket (or add a distinct bucket) on `MobileActionTargetProbe.StripsBindingOnMissing(r.TargetKind)` rather than solely on `r.BindingRemoved`, so a `KindWebPage` entry that could not be removed on the merge path still gets the "definitely dead, repoint it" message instead of the read-based "let the user decide" message.

- Major - Real-page regression test nulls ModelConfig, so the AC-1 primary-object exemption is never exercised against real Leads_FormPage data (`clio.tests/Command/McpServer/Tools/MobilePageConverter/WebToMobileRealPageRegressionTests.cs:350`). `CollectSourceEntityNames`/`ExceptSourceEntityTargets` exists specifically to stop the page's own primary object (Lead) from being falsely reported as missing a default mobile page, per its own doc comment noting the real Leads_FormPage declares twelve data sources across nine objects and that exempting all of them silenced the feature on the very page it was built for. The regression test that runs the probe against the pinned real Leads_FormPage fixture sets `ModelConfig = new JsonObject()` (empty), so `modelConfig?["dataSources"]` is never populated and the exemption path is never taken. In production wiring, `pageResponse.Bundle?.ModelConfig` is the real populated config, so the exemption WILL fire for Lead in practice, a materially different path than the one this regression test exercises. The one test built specifically to catch a regression on this exact scenario cannot detect a bug in the exemption logic. Fix: populate ModelConfig in the real-page regression test from the fixture's actual modelConfig (or a representative dataSources block naming Lead as the primary data source), and add an assertion that Lead's own create/update-record targets are exempted (absent from unresolvedTargetRequests) while LeadProduct/Opportunity are still reported.

- Minor - `ClassifyRelatedPageMetadata` treats a non-blank body without a well-formed "Pages" array as definitively Missing rather than Unknown (`clio/Command/McpServer/Tools/MobilePageConverter/MobileActionTargetProbe.cs:458`). The method's documented contract distinguishes a blank body (Unknown, no payload at all) from a well-formed body with an empty Pages array (Missing, a real observed answer). But a non-blank, parsed JObject that simply lacks a "Pages" property (`obj["Pages"] is not JsonArray`) also returns Missing, collapsing "the add-on genuinely declares no pages" and "the response shape wasn't what this probe expects" into the same definitive verdict. Fix: return `ActionTargetState.Unknown` (not Missing) when the parsed body lacks a recognizable "Pages" property, reserving Missing for a well-formed body whose Pages array is genuinely empty.

- Minor - Resolved/Missing/Unknown classification is tested by direct calls to an internal method, bypassing the public `Probe()` entry point (`clio/Command/McpServer/Tools/MobilePageConverter/MobileActionTargetProbe.cs`). The method's own catch-block comment notes it is internal and directly unit-tested, so a second caller could reach it without the outer degradation the probe's public path provides. Fix: route the Resolved/Missing/Unknown classification-matrix tests through `MobileActionTargetProbe.Probe()` with a substituted `IAddonSchemaDesignerClient.GetSchema` returning each metadata shape, keeping at most one or two direct calls only for a pure-parsing edge case not otherwise reachable.

- Minor - Sandbox E2E test can pass vacuously if no qualifying seeded page is found (`clio.mcp.e2e/MobilePageConversionGuideSandboxE2ETests.cs`). The test converts seeded pages "until one fires a navigating request" and then asserts on that finding's shape, but there is no explicit assertion that a qualifying page/finding was actually located before asserting on its properties. If sandbox seed data drifts so no page qualifies, the test can pass without exercising the AC-3 assertions at all. Fix: add an explicit assertion (e.g. `Should().NotBeEmpty()`/`NotBeNull()` on the located qualifying page or finding) before asserting on state/targetKind/elementName, so seed-data drift fails the test loudly instead of silently skipping the check.

**Policy line**

Suppressed by review policy: 3 outside the task's acceptance criteria; 1 already raised by another reviewer; 3 non-high-confidence Major(s) kept out of inline comments.

**Review by dkorotin1 — APPROVED** [review]
